libtenzir/src/tql2/eval_binary.cpp

```diff
@@ -191,6 +192,92 @@ struct BinOpKernel<ast::binary_op::add, duration_type, time_type> {
   }
 };
 
+template <>
+struct BinOpKernel<ast::binary_op::add, duration_type, duration_type> {
+  using result = duration;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_add(l.count(), r.count()); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration addition overflowed";
+  }
+};
+
+template <>
+struct BinOpKernel<ast::binary_op::sub, duration_type, duration_type> {
+  using result = duration;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_sub(l.count(), r.count()); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration subtraction overflowed";
+  }
+};
+
+template <>
+struct BinOpKernel<ast::binary_op::div, duration_type, duration_type> {
+  using result = double;
+
+  static auto evaluate(duration l, duration r)
+    -> std::variant<result, const char*> {
+    if (r == decltype(r){}) 
+      return "division by zero";
+    return detail::narrow_cast<result>(l.count()) / detail::narrow_cast<result>(r.count());
+  }
+};
+
+template <integral_type N>
+struct BinOpKernel<ast::binary_op::mul, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, type_to_data_t<N> r)
+    -> std::variant<result, const char*> {
+    if (auto check = checked_mul(l.count(), r); check.has_value()) {
+      return duration{check.value()};
+    }
+    return "Duration multiplication overflowed";
+  }
+};
+
+template<numeric_type N> requires std::same_as<double, type_to_data_t<N>>
+struct BinOpKernel<ast::binary_op::mul, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, double r)
+    -> std::variant<result, const char*> {
+    return duration_cast<duration>(l * r);
+  }
+};
+
+template <numeric_type N>
+struct BinOpKernel<ast::binary_op::mul, N, duration_type> {
+  using result = duration;
+
+  static auto evaluate(type_to_data_t<N> l, duration r)
+    -> std::variant<result, const char*> {
+    return BinOpKernel<ast::binary_op::mul, duration_type, N>::evaluate(r, l); // Commutative
+  }
+};
+
+template <numeric_type N>
+struct BinOpKernel<ast::binary_op::div, duration_type, N> {
+  using result = duration;
+
+  static auto evaluate(duration l, type_to_data_t<N> r)
+    -> std::variant<result, const char*> {
+    if (r == decltype(r){}) 
+      return "division by zero";
+
+    // NOTE: Might lose precision as the div op returns
+    // common type (aka double) but we use long as rep.
```

Comment:
This comment had me a bit confused, as you don't know that the common type is double here. You loose precision in any case, but for integers it happens with the division, and for doubles it happens only later, with the cast. The double truncation also applies to multiplication above, but as you discussed with Dominik, it probably doesn't matter. 

Maybe just remove this comment (and the blank line above it – no blank lines inside function definitions).

Reply:
Removed.